src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,169 @@
+        private readonly Decoder _decoder;
+        private readonly Encoder _encoder;
+        private char[] _charBuffer;
+        private int _charsDecoded;
+        private bool _disposed;
+
+        public TranscodingWriteStream(Stream stream, Encoding targetEncoding)
+        {
+            _stream = stream;
+
+            _charBuffer = ArrayPool<char>.Shared.Rent(MaxCharBufferSize);
+
+            // Attempt to allocate a byte buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the char -> byte conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            _maxByteBufferSize = Math.Min(MaxByteBufferSize, targetEncoding.GetMaxByteCount(MaxCharBufferSize));
+
+            _decoder = Encoding.UTF8.GetDecoder();
+            _encoder = targetEncoding.GetEncoder();
+        }
+
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+
+            ArraySegment<byte> bufferSegment = new ArraySegment<byte>(buffer, offset, count);
+            return WriteAsyncCore(bufferSegment, cancellationToken);
+        }
+
+        private async Task WriteAsyncCore(ArraySegment<byte> bufferSegment, CancellationToken cancellationToken)
```

Comment:
Is there a reason you chose to use `ArraySegment<byte>` rather than `ReadOnlyMemory<byte>`?  Note that JsonSerializer.Serialize ends up using the Stream.WriteAsync method that takes a `ReadOnlyMemory<byte>`, so it'd be more efficient if this stream overrode that WriteAsync method and made it the primary workhorse... it can then override the byte[]-based version to just delegate to the memory-based one.

Reply:
`WriteAsync(ReadOnlyMemory<Byte>, ...` is not available on NS2.0, but as per the Blazor team, they are fine with this package targeting NS2.1, is that change worth the upgrade?

Reply:
For the OOB, it's probably not worth it.  But it is for the netcoreapp build:
https://github.com/dotnet/runtime/pull/33459/files#diff-8d366f561ace3bceddbe061546d673f6R3

Reply:
On top of that, `ArraySegment` is what the original implementation in aspnetcore uses.